R/set_chat_client.R

```diff
@@ -0,0 +1,301 @@
+#'   set_chat_client("chat", chat_client, bookmark = "auto")
+#'
+#'   observeEvent(input$chat_user_input, {
+#'     stream <- chat_client$stream_async(input$chat_user_input)
+#'     chat_append("chat", stream)
+#'   })
+#' }
+#'
+#' # Enable bookmarking!
+#' shinyApp(ui, server, enableBookmarking = "url")
+#' @export
+set_chat_client <- function(
+  id,
+  client,
+  ...,
+  bookmark = c("auto", "manual"), # If more parameters are added, a third value of `"none"` can be used
+  session = getDefaultReactiveDomain()
+) {
+  rlang::check_dots_empty()
+  stopifnot(is.character(id) && length(id) == 1)
+
+  rlang::check_installed("ellmer")
+  if (!(inherits(client, "R6") && inherits(client, "Chat"))) {
+    rlang::abort(
+      "`client` must be an `ellmer::Chat()` object. If you would like to have {shinychat} support your own package, please submit a GitHub Issue at https://github.com/posit-dev/shinychat"
+    )
+  }
+  if (is.null(session)) {
+    rlang::abort(
+      "A `session` must be provided. Be sure to call `set_chat_client()` where a session context is available."
+    )
+  }
+
+  # if (isTRUE(bookmark)) {
+  set_chat_client_bookmark(
+    id,
+    client,
+    auto_update = match.arg(bookmark, several.ok = FALSE) == "auto",
+    session = session
+  )
+
+  # Don't return anything, even by chance
+  invisible(NULL)
+}
+
+#' @importFrom rlang %||%
+set_chat_client_bookmark <- function(
+  id,
+  client,
+  ...,
+  auto_update = TRUE,
+  session = getDefaultReactiveDomain()
+) {
+  rlang::check_dots_empty()
+
+  stopifnot(is.character(id) && length(id) == 1)
+
+  # Verify bookmark store is not "none"
+  bookmarkStore <- shiny::getShinyOption("bookmarkStore", "none")
+  if (bookmarkStore == "none") {
+    rlang::abort(
+      paste0(
+        "Error: Shiny bookmarking is not enabled. ",
+        "Please enable bookmarking in your Shiny app either by calling ",
+        "`shiny::enableBookmarking()` or by setting the parameter in ",
+        "`shiny::shinyApp(enableBookmarking = \"url\")`"
+      )
+    )
+  }
+
+  # Exclude works with bookmark names
+  excluded_names <- session$getBookmarkExclude()
+  ns_user_input <- paste0(id, "_user_input")
+  if (!(ns_user_input %in% excluded_names)) {
+    session$setBookmarkExclude(c(excluded_names, ns_user_input))
+  }
+
+  # Save
+  cancel_on_bookmark <- session$onBookmark(function(state) {
+    if (id %in% names(state$values)) {
+      rlang::abort(
+        paste0(
+          "Bookmark value with id (`\"",
+          id,
+          "\"`)) already exists. Please remove it or use a different id."
+        )
+      )
+    }
+
+    turns <- client$get_turns()
+    if (length(turns) == 0) return()
+
+    turns_list <- lapply(turns, function(turn) {
+      list(
+        role = turn@role,
+        # Convert everything to a single markdown string (including images!)
+        contents = ellmer::contents_markdown(turn),
+        tokens = turn@tokens
+      )
+    })
+    state$values[[id]] <- turns_list
```

Comment:
Can we just save the whole `client` object verbatim when `bookmarkStore` is `"server"`?

Also, I think we'll need that approach you had to serializing R6 objects for the `contents` --  `contents_markdown()` will be too lossy

Reply:
Yup! Good call!

Reply:
When doing URL formatting, contents_markdown() is the easiest way to get from text, image, or any other type to something that can be serialized.

I'd like to stay with `contents_markdown()` for now when doing URL formatting.